#' Amplification Input Table
#' 
#' This function calculates sensitivity outcomes for a specified range of DNA inputs.
#' @param copies_per_ng Copies of DNA per ng. Defaults to human genome (290).
#' @param total_primer_plexes Total plexes across which input DNA is divided.
#' @param loss_term Decimal proportion of total DNA lost during library preparation.
#' @param vaf_threshold Decimal minimum variant allele frequency threshold.
#' @param expected_vaf Decimal expected variant allele frequency. Defaults to germline heterozygous (0.5).
#' @param minimum_dna_input Minimum ng of input DNA to calculate probability matrix. Defaults to 0.
#' @param maximum_dna_input Maximum ng of input DNA to calculate probability matrix. Defaults to 3000.

amplification_input_table <- function(copies_per_ng=290, 
                                total_primer_plexes, 
                                loss_term, 
                                vaf_threshold, 
                                expected_vaf,
                                minimum_dna_input=0,
                                maximum_dna_input=3000,
                                ...){
  plexing_prob = 1/total_primer_plexes
  retained_template = 1-loss_term
  overall_prob = plexing_prob*retained_template # combined probabilities
  pre_amp_odds <- data.frame(dna_input=minimum_dna_input:maximum_dna_input)
  pre_amp_odds$prob=1-pbinom(ceiling(pre_amp_odds$dna_input*copies_per_ng*overall_prob*vaf_threshold), 
                             ceiling(pre_amp_odds$dna_input*copies_per_ng*overall_prob), 
                             prob=expected_vaf)
  return(pre_amp_odds)
}

#' Calculate Minimum Input
#' 
#' This function returns a minimum DNA input to achieve desired sensitivity.
#' @param pre_amp_odds Probability matrix of amplification odds. Can be generated by amplification_input_table.
#' @param sensitivity_goal Decimal sensitivity goal for the target panel.

calculate_minimum_input <- function(pre_amp_odds,
                          sensitivity_goal, ...){
  minimum_input <- min(dplyr::filter(pre_amp_odds, prob > sensitivity_goal)$dna_input)
  return(minimum_input)
}
  
#' Amplification Input Plot
#' 
#' This is the core plot function for comparing minimal input to current input.
#' @param pre_amp_odds Probability matrix of amplification odds. Can be generated by amplification_input_table.
#' @param amplification_input Total ng of DNA input to amplification.
#' @param minimum_input Minimum ng of DNA input for sensitivity goal. Can be generated by calculate_minimum_input.
#' @param sensitivity_goal Decimal sensitivity goal for the target panel.

amplification_input_plot <- function(pre_amp_odds, amplification_input, minimum_input, sensitivity_goal, ...){  
  ggplot(data=pre_amp_odds, aes(x=dna_input,y=prob)) + geom_line() +
    theme_bw() +
    scale_x_log10("Amplification Input (ng)", limits=c(1,3000), breaks=c(100, 200, 500, 1000, 2000)) +
    theme(axis.text.x = element_text(angle=90, hjust=1)) +
    geom_vline(xintercept=minimum_input, linetype=2) +
    geom_vline(xintercept=amplification_input) +
    annotate("text", label=paste('Current', as.character(
      amplification_input)), x=amplification_input, y=0.5, angle=90, size=10) +
    theme(text = element_text(size=24)) +
    ggtitle(paste('Minimum Input for ', sensitivity_goal*100, '% Sensitivity', sep='')) +
    annotate("text", label=paste('Minimum', as.character(minimum_input)), x=minimum_input, y=0.5, angle=90, size=10) +
    theme(axis.text = element_text(colour = "black")) +
    scale_y_continuous("Sensitivity")
  }

#' Plot Minimal Input
#' 
#' Master plotting function that combines all package utilities to generate a 
#' comprehensive plot of minimum vs. current DNA inputs.
#' @param copies_per_ng Copies of DNA per ng. Defaults to human genome (290).
#' @param total_primer_plexes Total plexes across which input DNA is divided.
#' @param loss_term Decimal proportion of total DNA lost during library preparation.
#' @param amplification_input Total ng of DNA input to amplification.
#' @param vaf_threshold Decimal minimum variant allele frequency threshold.
#' @param expected_vaf Decimal expected variant allele frequency. Defaults to germline heterozygous (0.5).
#' @param minimum_dna_input Minimum ng of input DNA to calculate probability matrix. Defaults to 0.
#' @param maximum_dna_input Maximum ng of input DNA to calculate probability matrix. Defaults to 3000.
#' @param sensitivity_goal Decimal sensitivity goal for the target panel. Defaults to 0.999 (99.9%).

plot_minimal_input <- function(copies_per_ng=290, 
                        total_primer_plexes, 
                        loss_term, 
                        amplification_input, 
                        vaf_threshold, 
                        expected_vaf,
                        minimum_dna_input=0,
                        maximum_dna_input=3000,
                        sensitivity_goal=0.999){
  params <- c(as.list(environment()))
  pre_amp_odds <- do.call(amplification_input_table, params)
  minimum_input <- calculate_minimum_input(pre_amp_odds, sensitivity_goal)
  amplification_input_plot(pre_amp_odds, amplification_input, minimum_input, sensitivity_goal)
}
