#' Nested Read Depth Plot
#' 
#' This is the core plot function for comparing read depth to the nested binomial sensitivity.
#' @param read_depth_table Probability matrix of read depths. Can be generated by read_depth_table().
#' @param sensitivity_goal Decimal sensitivity goal for the target panel.
#' @param required_depth Minimum read depth requirement for sensitivity goal. Can be generated by calculate_required_depth().
#' @param minimum_depth Minimum read depth to calculate probability matrix. Defaults to 0.
#' @param maximum_depth Maximum read depth to calculate probability matrix. Defaults to 3000.

nested_read_depth_plot = function(
  read_depth_table, sensitivity_goal, required_depth, minimum_depth=10, maximum_depth=3000, ...){
  
  ggplot(data=read_depth_table, aes(x=read_depth, y=prob)) + 
    geom_smooth(method='lm', formula = y ~ poly(x, 5)) +
    theme_bw() +
    theme(text = element_text(size=20)) +
    theme(axis.text.x = element_text(angle=90, hjust=1)) +
    ggtitle("Nested Binomial Sensitivity at Read Depth") +
    geom_hline(yintercept=sensitivity_goal, linetype=2) +
    geom_vline(xintercept=required_depth, linetype=2) +
    annotate("text", label=paste('Minimum', as.character(required_depth)), x=required_depth, y=0.95, angle=90, size=10) +
    scale_y_continuous("Sensitivity") +
    coord_cartesian(ylim=c(0.9,1)) +
    scale_x_log10("Paired Read Depth", limits=c(minimum_depth, max(c(read_depth_table$read_depth, required_depth))), 
                  breaks=c(10, 20, 50, 100, 200, 500, 1000, minimum_depth)) 
}

#' Nested Copy Calculation
#' 
#' This function calculates the fraction of original genomic copies expected to appear in N reads.
#' @param total_primer_plexes Total plexes across which input DNA is divided.
#' @param minimum_depth Minimum read depth to calculate probability matrix. Defaults to 0.
#' @param maximum_depth Maximum read depth to calculate probability matrix. Defaults to 3000.
#' @param loss_term Decimal proportion of total DNA lost during library preparation.
#' @param dna_input Total ng of input DNA entering amplification reaction.
#' @param expected_vaf Decimal expected variant allele frequency.
#' @param vaf_threshold Decimal minimum variant allele frequency threshold.
#' @param min_alt_reads Minimum alternate supporting reads to call the variant. Defaults to 2.
#' @param copies_per_ng Copies of DNA per ng. Defaults to human genome (290).

nested_read_depth_table <- function(total_primer_plexes,
                                    minimum_depth=0,
                                    maximum_depth=3000,
                                    loss_term,
                                    dna_input,
                                    expected_vaf,
                                    vaf_threshold,
                                    min_alt_reads=2,
                                    copies_per_ng=290,
                                    ...){
  plexing_prob = 1/total_primer_plexes
  retained_template = 1-loss_term
  overall_prob = plexing_prob*retained_template # combined probabilities
  starting_genomic_copies = round(overall_prob*copies_per_ng*dna_input)
  if(starting_genomic_copies > 0){
    nested_odds <- data.frame(read_depth=minimum_depth:maximum_depth)
    nested_odds$odds <- (1-((starting_genomic_copies-1)/starting_genomic_copies)^nested_odds$read_depth)
    nested_odds$final_genomic_copies <- floor(nested_odds$odds*starting_genomic_copies)
    nested_odds$min_alt_reads <- ceiling(nested_odds$final_genomic_copies*vaf_threshold)
    nested_odds$min_alt_reads[(nested_odds$min_alt_reads <= min_alt_reads)] <- min_alt_reads
    nested_odds$prob=1-pbinom(nested_odds$min_alt_reads, 
                              nested_odds$final_genomic_copies, expected_vaf)
    return(dplyr::select(nested_odds, read_depth, prob))}
  else {
    return('Error: <1 starting genomic copies.')}
}

#' Plot Minimal Nested Binomial Depth
#' 
#' Master plotting function that combines all package utilities to generate a 
#' plot of minimum vs. current read depths.
#' @param vaf_threshold Decimal minimum variant allele frequency threshold.
#' @param total_primer_plexes Total plexes across which input DNA is divided.
#' @param dna_input DNA amount added to targeted PCR.
#' @param copies_per_ng Copies of DNA per ng. Defaults to human genome (290).
#' @param expected_vaf Decimal expected variant allele frequency. Defaults to germline heterozygous (0.5).
#' @param loss_term Decimal proportion of total DNA lost during library preparation.
#' @param minimum_depth Minimum read depth to calculate probability matrix. Defaults to 0.
#' @param maximum_depth Maximum read depth to calculate probability matrix. Defaults to 3000.
#' @param min_alt_reads Minimum alternate supporting reads to call the variant. Defaults to 2.
#' @param sensitivity_goal Decimal sensitivity goal for the target panel. Defaults to 0.999 (99.9%).

plot_nested_minimal_depth <- function(vaf_threshold, 
                               total_primer_plexes,
                               dna_input,
                               copies_per_ng=290,
                               expected_vaf=0.5,
                               loss_term=0,
                               minimum_depth=1,
                               maximum_depth=3000,
                               min_alt_reads=2,
                               sensitivity_goal=0.999){
  params <- c(as.list(environment()))
  read_depth_table <- do.call(nested_read_depth_table, params)
  required_depth <- calculate_required_depth(read_depth_table, sensitivity_goal)
  nested_read_depth_plot(read_depth_table, sensitivity_goal, current_depth, required_depth)
}
